WIP: Automatically Persist Filter Value on Blur

marcushyett-ph opened the pull request:
This is a work in progress - I'd like to get feedback on my approach as I'm not super confident it's correct. 

The following video shows adding a new value to a multi-value-field (equals) by: 
* Selecting a value
* Clicking outside of the edit area
* Pressing enter
* Pressing escape (where the most recent input is expect to be lost)

It also shows the same scenarios for a single value field (Contains)

https://user-images.githubusercontent.com/85295485/127501346-f997aa3b-a5a4-4a39-8e99-ad18d0f4d09e.mov

## Changes

* Added a handler to automatically persist the value of the input on blur
* Resolved bug where suggested value was a comma separated list of existing values
* Preserved escape key functionality to empty and blur the field

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests
- [ ] Migrations are safe to run at scale (e.g. PostHog Cloud) – present proof if not obvious
- [ ] New/changed UI is decent on smartphones (viewport width around 360px)

MarconLP committed:
Automatically Persist Filter Value on Blur

* Added a handler to automatically persist the value of the input on blur
* Resolved bug where suggested value was a comma separated list of existing values
* Preserved escape key functionality to empty and blur the field

marcushyett-ph commented on frontend/src/lib/components/PropertyFilters/components/PropertyValue.tsx:
```diff
@@ -70,13 +70,6 @@ export function PropertyValue({
     useEffect(() => {
         if (!value) {
             setInput('')
-        } else if (value !== input) {
```
I genuinely cannot work out what value this code has - I've removed it and tried a number of scenarios and cannot see any adverse effects - it'd be amazing if @alexkim205 you might be able to share any context on why this might be here.

I've removed it because it causes this bug in production where suggest an option which is equivalent to a concatenated string of all the existing values.

![image](https://user-images.githubusercontent.com/85295485/127502332-4fef37c0-7064-431c-b733-264e2e4f2b46.png)

paolodamico replied:
Tested too and couldn't find any issues, I think it should be safe to remove. Tagging @samwinslow in case you have any context too

marcushyett-ph replied:
It looks like it might be trying to do something if the values are JSON objects - but I can't think why the values would be JSON objects - perhaps I'm missing something?

alexkim205 replied:
Looked at the code briefly and I think removing this JSON check might break functionality in use cases where we pass in a `value` property that represents a filter id. For example, `<PropertyValue/>` is still being used like this in [Path Tab](https://github.com/PostHog/posthog/blob/7667d128e815bf6f135a029c52c2298633f4dc4a/frontend/src/scenes/insights/InsightTabs/PathTab.tsx#L59-L59) where we're passing a numeric `filter.start_point` in as `value` which will always trigger this code path. This will set the input text as a numeric value on refresh/`value` change, instead of the name of the property associated with that numeric id.

I don't think we should remove this until we fully deprecate the other consumers of this component, or move this logic one up the component tree.

paolodamico replied:
Great catch! Let's bring back this code!

marcushyett-ph commented on frontend/src/lib/components/SelectGradientOverflow.tsx:
```diff
@@ -46,6 +46,7 @@ export type SelectGradientOverflowProps = SelectProps<any> & {
     delayBeforeAutoOpen?: number
     dropdownMatchSelectWidth?: boolean | number
     placement?: 'bottomLeft' | 'topLeft' // Dropdown placement (undefined = auto). See API at https://ant.design/components/dropdown
+    handleBlur?: () => void
```
I get a curious React error which tells me that handleBlur should be lowercase

![image](https://user-images.githubusercontent.com/85295485/127502621-d5d19b21-a906-4a54-a87f-da02ca513874.png)

But I'm not sure if I've done something more fundamentally wrong - so didn't want to just change it to be lowercase

paolodamico replied:
Oh I see what the problem is, `handleBlur` was being passed down all the way to the raw HTML component through `props` because it wasn't destructured. See the commit I pushed, we extract the `handleBlur` property so it doesn't get rendered in the downstream components.

marcushyett-ph replied:
Ah nice catch.

paolodamico commented on frontend/src/lib/components/PropertyFilters/components/PropertyValue.tsx:
```diff
@@ -120,6 +113,12 @@ export function PropertyValue({
     }, [propertyKey])
 
+    useEffect(() => {
+        if (input === '' && document.activeElement instanceof HTMLElement) {
```
This whole `useEffect` I'm not sure about, if you just want to erase everything to type something else, losing focus on the input seems annoying. Curious on the context for this.

marcushyett-ph replied:
The context is a little complex so bare with me, would love any suggestions for a better way to do this:

1) Since I'm now handling the blur event in a different function, the previous code (when your press Esc key) would trigger a blur - invoking the new blur logic - which stipulates that we should persist the current value - this is not what the user would expect

2) Since the handleBlur function does not persist the value if the input is empty (which it should be after you've pressed escape), I would like to set the input value to empty, then trigger the handleBlur function, such that on escape the blur occurs after the input value has been set to empty and the value is therefore not persisted (expected behavior)

3) However, setState takes time to propagate the value and blur does not. Ideally I would've used a callback on the `setInput('')` call, however since we're using `useState`, callbacks are not supported (I think they actually work, but they're not supported officially). So I can't see an elegant way to execute the blur once the `setInput('')` call has updated the state successfully.

4) I tried another hack with a `setTimeout` but it felt mega-ugly

5) Yes, the problem is that if someone backspaces the box empty - we'll lose focus. I don't think this is a major issue - it's smaller than the existing issue this PR is designed to fix - but I would rather it didn't happen if you can think of another nice solution?

6) The only other option I considered was to refactor `setInput()` to use `setState()` rather than `useState()` and add a callback on the setState to trigger the blur - this felt like it would add unnecessary complexity - but might be the best solution

alexkim205 replied:
It's great that you considered all these other options. I agree w/ Paolo that it could be annoying on input blur everytime input is empty and I think we can kill both birds w/ one stone here. If I'm understanding correctly, are you trying to blur the text field only when a user presses the escape key (and not when input becomes empty via backspace)? If that's the case we can react to a separate `shouldBlur` state. Ugly but still a lot less overhead than most other options.

alexkim205 replied:
Demo

https://user-images.githubusercontent.com/13460330/127895740-39d88a79-72ef-49c7-968d-e7ca6196cc10.mov

Pushed a few changes to implement this.

marcushyett-ph replied:
Amazing thanks @alexkim205  really appreciate both your feedback and changes.

@paolodamico since @alexkim205  has made changes here since you originally approved - are you okay with us to merge this latest version?

paolodamico committed:
fix handleBlur prop

paolodamico committed:
fix a separate bug

paolodamico reviewed (approved):
changes lgtm! fixed the React bug and added a few comments inline to review before merging.

alexkim205 commented on frontend/src/lib/components/PropertyFilters/components/PropertyValue.tsx:
```diff
@@ -154,6 +153,16 @@ export function PropertyValue({
                 }
             }
         },
+        handleBlur: () => {
```
nice! like that this is a prop rather than in a useEffect

alexkim205 committed:
make blurring happen after input is set to empty

paolodamico committed:
simplify blur on escape

alexkim205 commented on frontend/src/lib/components/SelectGradientOverflow.tsx:
```diff
@@ -89,6 +91,7 @@ export function SelectGradientOverflow({
     const onBlur = (): void => {
         if (isOpen) {
             setOpen(false)
+            handleBlur?.()
```
nit but prefer this to have a default value of `()=>{}` rather than `nullish ?` syntax

paolodamico commented:
Oops! I just saw your working on this @alexkim205, I'll let you take it on

paolodamico commented on frontend/src/lib/components/PropertyFilters/components/PropertyValue.tsx:
```diff
@@ -142,7 +143,8 @@ export function PropertyValue({
             if (e.key === 'Escape' && e.target instanceof HTMLElement) {
-                e.target.blur()
+                setInput('')
+                setShouldBlur(true)
```
I actually think this is dirtier than what we had before. How about just calling `e.target.blur()` directly here and getting rid of the entire `useEffect` above (lines 117 - 123)? @marcushyett-ph

marcushyett-ph replied:
If we do that when you hit escape we will persist the value you were typing rather than clearing the input?

paolodamico reviewed (changes requested):
Added two outstanding comments that I think are worth addressing before merging

MarconLP committed:
Update PropertyValue.tsx

Re-aded elseif around valueObject - however removed the else if it is not a value object since this causes a bug in the autocomplete

MarconLP committed:
Merge branch 'marcushyett-auto-save-filter-value-on-blur' of https://github.com/PostHog/posthog into marcushyett-auto-save-filter-value-on-blur

paolodamico committed:
minor cleanup
